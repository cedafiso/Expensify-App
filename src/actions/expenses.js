import { ref, remove, set, update } from '@firebase/database';
import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';
import { firebase } from '../firebase/firebase';


export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0, id = uuidv4()} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id,
        description,
        note,
        amount,
        createdAt
    }
});
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', note = '', amount = 0, createdAt = 0
        } = expenseData;
        const expense = {description, note,amount,createdAt}
        firebase.push(ref(database, 'expenses'), expense).then((ref) => {
            console.log(ref.key);
            dispatch(addExpense(
                Object.assign(expense, {id: ref.key})
            ));
        });
    }
};
export const removeExpense = ({ id }={}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = (expenseData) => {
    return (dispatch) => {
        const {id} = expenseData;
        remove(ref(database, `expenses/${id}`)).
        then(() => {
            console.log('Sucefully remove');
        }).
        catch((e) =>{
            console.log('Unsucefully remove');
        });
        dispatch(removeExpense({id}));
    }
};

export const startEditExpense = (id, expense) => {
    return (dispatch) => {
        const {
            description = '', note = '', amount = 0, createdAt = 0
        } = expense;
        const updates = {description, note, amount, createdAt};
        update(ref(database, 'expenses/'+ id), updates);
        dispatch(editExpense(id, expense));
    };

};

export const editExpense = (id, updates) => ({
    type: 'EDIT',
    id,
    updates
});

export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses  
});

export const startSetExpenses = () => {
    return (dispatch) => {
        const expense = [];
        firebase.get(firebase.child(firebase.ref(database),'expenses')).
        then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expense.push(Object.assign(childSnapshot.val(), {id:childSnapshot.key}));
            })
            dispatch(setExpenses(expense));
        })
    }
};
