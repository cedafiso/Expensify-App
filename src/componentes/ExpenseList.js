import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import visibleExpenses from '../selectors/expenses';


export const ExpenseList = (props) => {
    return (
        <div>
            <h1>Expense List:</h1>
            {
                props.expenses.length === 0 && <p>Empty</p>
            }
            {props.expenses.map((item) => {
                return (<ExpenseListItem
                    key={item.id}
                    {...item}

                />)
            })}
        </div>
    )
};

const ConnectedExpenseList = connect()

const mapStateToProps = (state) => {
    return {
        expenses : visibleExpenses(state.expenses , state.filters)
    };
}

export default connect(mapStateToProps)(ExpenseList);