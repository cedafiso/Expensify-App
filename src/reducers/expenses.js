const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state,action.expense
            ]
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => { return id !== action.id})
        case 'EDIT':
            return state.map((expense) =>{
                if (expense.id === action.id){
                    return (Object.assign({}, expense, action.updates));
                }else{
                    return expense
                }
            })
        case 'SET_EXPENSES':
            return state.concat(action.expenses)
        default:
            return state;
    }
};

export default expensesReducer;