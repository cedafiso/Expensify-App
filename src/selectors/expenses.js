import moment from 'moment';

export default (expense, { endDate , sortBy = '', startDate, text =''} = {}) => {
    return expense.filter((expense) => {
        const creatAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(creatAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(creatAtMoment, 'day') : true;
        const textMatch = text === '' || expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=> {
        if( sortBy === 'date'){
            return (a.createdAt < b.createdAt ? 1 : -1)
        } else if (sortBy === 'amount'){
            return (a.amount < b.amount ? 1 : -1)
        }
    });
};