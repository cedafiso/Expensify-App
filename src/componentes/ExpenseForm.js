import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker} from 'react-dates';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';


export default class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense ? this.props.expense.description : '',
        amount: this.props.expense  ? (this.props.expense.amount/100).toString() : '',
        note: this.props.expense  ? this.props.expense.note : '',
        createdAt: this.props.expense  ? moment(this.props.expense.createdAt) : moment(),
        focused: false,
        error:'',
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}));
        }else{
            return null;
        }
        
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}));
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(() => ({focused}));
    }
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.amount || !this.state.description){
            this.setState(() => ({error: 'Please provide description and amount'}));
        }else{
           this.setState(() => {error:''});
           this.props.onSubmit({
               description: this.state.description,
               amount: parseFloat(this.state.amount, 10)*100 ,
               createdAt: this.state.createdAt.valueOf(),
               note: this.state.note
           });
          }
    };
    render(){
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.error && <p>{this.state.error}</p>}
                    <input 
                    type='text'
                    placerholder='Description'
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    autoFocus/>
                    <input 
                    type='text'
                    placeholder='Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}

                    />
                    <SingleDatePicker
                    date ={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.focused}
                    onFocusChange={this.onFocusChange}
                    />
                    <textarea 
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    />
                    <button>Submit</button>
                </form>
            </div>
        )
    }

}

connect()(ExpenseForm);