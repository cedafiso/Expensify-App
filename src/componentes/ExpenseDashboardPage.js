import React from 'react';
import ExpenseList from './ExpenseList';

const ExpenseDashboardPage = () => {
    return (
        <div>
            <p>Welcome</p>
            <ExpenseList/>
        </div>
    );
}

export default ExpenseDashboardPage;