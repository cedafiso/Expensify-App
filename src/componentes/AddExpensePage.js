import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {startAddExpense} from '../actions/expenses';

const AddExpensePage = (props) => {
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__tittle">Add expense</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm 
                onSubmit={(expenses) => {
                    props.dispatch(startAddExpense(expenses));
                    props.history.push('/');
                }}
                />
            </div>
        </div>
    );
}

export default connect()(AddExpensePage);