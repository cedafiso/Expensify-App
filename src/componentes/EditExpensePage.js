import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {editExpense} from '../actions/expenses';
import {useParams} from 'react-router-dom';

const EditExpensePage = (props) => {
    console.log(props.expenses);
    return (
        <div>
            <p>Expensify-App-Edit {props.match.params.id}</p>
            <ExpenseForm
                expense={props.expenses}
                onSubmit={(expenses) => {
                props.dispatch(editExpense(props.match.params.id, expenses));
                props.history.push('/');
            }
                }/>
        </div>
    );
}

const mapStateToProps = (state, props) => ({
    expenses: state.expenses.find((expense) => {
        return expense.id === props.match.params.id})
});

export default connect(mapStateToProps)(EditExpensePage);