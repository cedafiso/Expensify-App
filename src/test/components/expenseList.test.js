import React from 'react';
import {shallow} from 'enzyme';
import { ExpenseList } from "../../componentes/ExpenseList";
import {state} from '../fixtures/expenses';

test('Should render an instance of ExpenseList', () => {
    const wrapper = shallow(<ExpenseList expenses={state}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Should render an instance of ExpenseList with empty sign', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});