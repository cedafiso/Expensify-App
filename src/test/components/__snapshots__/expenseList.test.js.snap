// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render an instance of ExpenseList 1`] = `
<div>
  <h1>
    Expense List:
  </h1>
  <Connect(ExpenseListItem)
    amount={1000}
    createdAt={0}
    description="rent"
    id="1"
    key="1"
    note="apt"
  />
  <Connect(ExpenseListItem)
    amount={8000}
    createdAt={10000}
    description="Water bill"
    id="2"
    key="2"
    note="apt"
  />
  <Connect(ExpenseListItem)
    amount={500}
    createdAt={-10000}
    description="Gas bill"
    id="3"
    key="3"
    note="apt"
  />
</div>
`;

exports[`Should render an instance of ExpenseList with empty sign 1`] = `
<div>
  <h1>
    Expense List:
  </h1>
  <p>
    Empty
  </p>
</div>
`;
